File: OJ_Assignment/Ch02-Linear_List/1009.2.18.py
class ListNode:
    def __init__(self, data = None):
        self.val = data
        self.nxt = None

class LinkList:
    def __init__(self):
        self.head = None
    
    def print(self):
        p = self.head
        if p is None:
            print("NULL")
        else:
            while p is not None:
                if p.nxt is not None:
                    print(p.val, end=" ")
                else:
                    print(p.val)
                p = p.nxt
    
    def delete(self, index):
        p = self.head
        if index == 0:
            self.head = self.head.nxt
            del p
        else:
            cnt = 1
            while cnt < index:
                p = p.nxt
                cnt += 1
            q = p.nxt
            p.nxt = q.nxt
            del q
    
L = LinkList()
data = list(map(int, input().split()))
p = L.head = ListNode(data[0])
for i in range(1, len(data)):
    p.nxt = ListNode(data[i])
    p = p.nxt
tar = int(input())
L.delete(tar)
L.print()

File: OJ_Assignment/Ch02-Linear_List/1009.2.18.cpp
#include <iostream>

using namespace std;

typedef struct ListNode{
    int val;
    ListNode *nxt;

    ListNode(int e = 0, ListNode *ptr = nullptr) :
        val(e), nxt(ptr) {}
} *LinkList;

void PrintLinkList(LinkList p){
    if (p == nullptr) {
        printf("NULL\n");
        return;
    }
    while (p != nullptr) {
        printf("%d ", p->val);
        p = p->nxt;
    }
    putchar('\n');
}

int main() {
    LinkList head = nullptr;
    ListNode *p = nullptr;
    int tmp = 0, cnt = 0;

    cin >> tmp;
    p = head = new ListNode(tmp);
    while (cin >> tmp) {
        p->nxt = new ListNode(tmp, p);
        p = p->nxt;
    }

    ListNode *q = p->nxt;
    q->nxt = nullptr;
    delete p;
    
    p = head;
    if (tmp == 0) {
        head = head->nxt;
        delete p;
        p = head;
    } else {
        while (++cnt < tmp) {
            p = p->nxt;
        }
        ListNode *q = p->nxt;
        p->nxt = q->nxt;
        delete q;
        q = nullptr;
    }

    PrintLinkList(head);
    return 0;
}